### PR TITLE
[core][misc] improve free_finished_seq_groups

### PR DESCRIPTION
**Pull request by youkaichao**
`free_finished_seq_groups` is in the critical path, because the engine calls it after every model step.

the main branch searchs for all the requests to find finished requests, while new finished requests should only appear in running queue.

this simple optimization, turns out to be important, when we process large amount of offline batch inference, e.g. when users give `100k` prompts to the LLM engine. In this case, we are iterating all 100k data at every step.

I have heard users complaining that sending the whole dataset to `LLM` is slow, and he has to chunk the data into small batches (25 requests each) to make it fast. This PR might fix it.

Some micro benchmark to demonstrate the benefit when number of prompts are long and the model becomes small (thus overhead in scheduler becomes significant):

Note: main branch is commit https://github.com/vllm-project/vllm/commit/925de97e05dd4709fcd80691cb37da5e582c22e8

```shell
$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 10000

# main branch:
Throughput: 577.63 requests/s, 18484.31 tokens/s

# this PR:
Throughput: 674.46 requests/s, 21582.66 tokens/s

$ python benchmarks/benchmark_throughput.py --output-len 16 --input 16 --model facebook/opt-125m --num-prompts 100000

# main branch:
Throughput: 211.18 requests/s, 6757.85 tokens/s

# this PR:
Throughput: 522.91 requests/s, 16733.13 tokens/s
```

Conclusion:

When I increase the number of prompts from  10k to 100k, the main branch is slowed down by almost 3x. Improving the `free_finished_seq_groups` helps a lot, the slow down is about 40%. Apparently there are some other factors, but I think this might be the largest factor in processing with large datasets.

### CODE REVIEW

File: vllm/core/scheduler.py
Comment (youkaichao) [inline vllm/core/scheduler.py:1061]: I notice that this code is introduced by https://github.com/vllm-project/vllm/pull/4115 @mzusman . Please take a look to see if it makes sense to you.

**Comment by youkaichao** [comment]
Here is the flamechart of the main branch, with 100k prompts. It becomes clear that `free_finished_seq_groups` is the major source of overhead.

flamechart

**Comment by youkaichao** [comment]
Here is the flame chart of this PR, with various number of prompts:

10k prompts:

flamechart_10k

100k prompts:

flamechart_100k

500k prompts:
flamechart_500k

Even after this PR, there is still a slow down factor that is proportional to the number of waiting requests. The flamechat points to:

https://github.com/vllm-project/vllm/blob/fad5576c58864a6c2cf528f67e60e03a949b3dac/vllm/core/scheduler.py#L680 

which is indeed very inefficient.

**Review by mgoin — APPROVED** [review]
This makes a lot of sense, matches behavior I've seem or been reported to me, and should not be controversial. LGTM pending review by @mzusman

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:1061]: Yeah, Looks really good. We've missed that use-case. Thank you for fixing it up!

**Review by mzusman — APPROVED** [review]

**Comment by mgoin** [comment]
@youkaichao I think you must rebase to pass the failing Examples Test
```
[2024-07-29T19:01:27Z] python3: can't open file '/vllm-workspace/examples/llava_example.py': [Errno 2] No such file or directory
```

File: vllm/core/scheduler.py
Comment (njhill) [inline vllm/core/scheduler.py:1073]: @youkaichao how about wrapping this in
```python
 if any(seq_group.is_finished() for seq_group in self.running):
 ```
 to avoid rebuilding the deque every time, since it requests will be finished relatively rarely.

File: vllm/core/scheduler.py
Comment (WoosukKwon) [inline vllm/core/scheduler.py:1063]: @mzusman I actually don't get the current (before this PR) logic here.

If I understand correctly, the order of execution is `schedule` -> `get_and_reset_finished_requests_ids` -> model execution -> `free_finished_seq_groups`. Hence, the requests freed from `self.swapped` or `self.waiting` were not actually added to `finished_requests_ids` before the model execution. Is this the right order of execution?

Comment (WoosukKwon) [reply vllm/core/scheduler.py:1063]: Ok, to my understanding,

1. Actually `finished_requests_ids` doesn't have to include the requests rejected in `self.waiting` since they are never passed into the model runner.
2. `finished_requests_ids` must include the requests rejected in `self.swapped` in the current step, which means the current code (before this PR) has a bug. However, the bug has not been noticed because this case is very rare.

File: vllm/core/scheduler.py
Comment (youkaichao) [reply vllm/core/scheduler.py:1073]: I plan to refactor this into dictionary in the future, so that we can easily delete requests.

**Comment by youkaichao** [comment]
> @youkaichao I think you must rebase to pass the failing Examples Test
> 
> ```
> [2024-07-29T19:01:27Z] python3: can't open file '/vllm-workspace/examples/llava_example.py': [Errno 2] No such file or directory
> ```

If there's no more comments, I will force-merge to avoid one additional ci cost :)

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
@youkaichao I think we need to understand the code before getting this PR merged.

File: vllm/core/scheduler.py
Comment (WoosukKwon) [inline vllm/core/scheduler.py:703]: I believe this can be skipped because `finished_requests_ids` is to release the mamba cache of the already-scheduled requests while the ignored requests here are never scheduled.

File: vllm/core/scheduler.py
Comment (WoosukKwon) [inline vllm/core/scheduler.py:719]: ditto

**Comment by youkaichao** [comment]
> @youkaichao I think we need to understand the code before getting this PR merged.

LGTM. Please take your time. I think the scheduler logic is quite unclear though. Might have lots of bugs there.

**Comment by youkaichao** [comment]
@WoosukKwon please feel free to take this PR if you want some modifications!

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:1063]: Yeah, That's correct, `finished_requests_ids` holds information that is lagged by 1 step. finished requests will always be released upon the next step.

AFAIU, `self.waiting` also includes preempted requests that got rescheduled, preempted requests did previously passed into the model runner and are already registered in the mamba cache. If those requests get aborted then we do add them to the `finished_requests_ids` and release them through [here](https://github.com/vllm-project/vllm/blob/135d8fefbf4242d4bf8c1a8b9f670047adc13b61/vllm/core/scheduler.py#L377) though. 
BTW Just to be sure, by requests rejected, do you mean aborted requests?

So actually, yeah, checking for finished requests in the `self.waiting` and `self.swapped` in `free_finished_seq_groups` is not necessary.

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:703]: I agree.

File: vllm/core/scheduler.py
Comment (WoosukKwon) [reply vllm/core/scheduler.py:1063]: @mzusman Thanks for the explanation! However, it's a bit unclear to me how it handles the preemption, both recompuation and swapping. Could you elaborate more on that?

Comment (mzusman) [reply vllm/core/scheduler.py:1063]: @WoosukKwon Sure! Upon recomputation we actually keep the Mamba cache as it is and do not evict it (since Mamba cache is quite small), the request id persists during the preemption therefore we can still use it's corresponding cache upon recomputation. And upon swapping - We do not handle swapping at the moment as it occurs fairly rarely and it's quite complicated to implement atm.

Comment (WoosukKwon) [reply vllm/core/scheduler.py:1062]: @mzusman Thanks for the detailed explanation! It is super helpful to understand the implementation.

I think we should revisit this design decision in the near future. Currently, it's a bit confusing for code readers.

**Review by WoosukKwon — APPROVED** [review]
@youkaichao I've update the PR based on the discussion with @mzusman. I think the PR is good to go now.

**Pull request merged**
